File: 00.My_Parctice_Courses/03.Algorithms_Java_Solutions/04.Dynamic_Programming/DynamicProgrammingLAB/src/LIS/LongestIncreasingSubsequence.java
package LIS;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LongestIncreasingSubsequence {

    public static void main(String[] args) throws IOException {

        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
//        int[] sequence = Arrays.stream(bf.readLine().split(" ")).
//                mapToInt(Integer::parseInt).toArray();

        int[] sequence = {9,100,20,30,40};
        List<Integer> lis = findLIS(sequence);

        for (Integer num : lis) {
            System.out.print(num + " ");
        }
    }

    private static List<Integer> findLIS(int[] sequence) {
        int[] len = new int[sequence.length];
        int[] prev = new int[sequence.length];

        int lastIndex = -1;
        int maxLen = 0;

        for (int i = 0; i < sequence.length; i++) {

            int currentNum = sequence[i];
            len[i] = 1;
            prev[i] = -1;
            for (int j = 0; j < i; j++) {
                if (currentNum > sequence[j] && len[j] + 1 > len[i]) {
                    len[i] = len[j] + 1;
                    prev[i] = j;
                }
            }
            if (len[i] > maxLen) {
                maxLen = len[i];
                lastIndex = i;
            }
        }

        List<Integer> lis = recoverSubsequence(lastIndex, prev, sequence);
        return lis;
    }

    private static List<Integer> recoverSubsequence(int lastIndex, int[] prev, int[] sequence) {
        List<Integer> lis = new ArrayList<>();

        while (lastIndex != -1) {
            lis.add(sequence[lastIndex]);
            lastIndex = prev[lastIndex];
        }
        Collections.reverse(lis);
        return lis;
    }
}

File: 00.My_Practice_Courses/02Algorithms_April_2016/04.Dynamic_Programming_Lab/Longest-Increasing-Subsequence/LongestIncreasingSubsequence.cs
namespace Longest_Increasing_Subsequence
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class LongestIncreasingSubsequence
    {
        public static void Main()
        {
            var sequence = Console.ReadLine().
                Split().
                Select(int.Parse).
                ToArray(); //new[] { 3, 14, 5, 12, 15, 7, 8, 9, 11, 10, 1 };
            //longest sequence is 3 5 7 8 9 11
            var longestSeq = FindLongestIncreasingSubsequence(sequence);
            Console.WriteLine("Longest increasing subsequence (LIS)");
            Console.WriteLine("  Length: {0}", longestSeq.Length);
            Console.WriteLine("  Sequence: [{0}]", string.Join(", ", longestSeq));
        }

        public static int[] FindLongestIncreasingSubsequence(int[] sequence)
        {
            var len = new int[sequence.Length];
            var prev = new int[sequence.Length];

            int maxLen = 0;
            int lastIndex = -1;

            for (int x = 0; x < sequence.Length; x++)
            {
                var currentElement = sequence[x];
                len[x] = 1;
                prev[x] = -1;

                for (int i = 0; i < x; i++)
                {
                    if (sequence[i] < currentElement && len[i] >= len[x])
                    {
                        len[x]++;
                        prev[x] = i;
                    }
                }
                if (len[x] > maxLen)
                {
                    maxLen = len[x];
                    lastIndex = x; 
                }
            }
            var longestSeq = new List<int>();
            while (lastIndex != -1)
            {
                longestSeq.Add(sequence[lastIndex]);
                lastIndex = prev[lastIndex];
            }
            longestSeq.Reverse();
            return longestSeq.ToArray();
        }
    }
}
